source/UF/ConnectedTypes.lagda

```diff
@@ -0,0 +1,223 @@
+
+ connectedness-extends-below : {X : 𝓤 ̇} {k l : ℕ}
+                             → (l ≤ℕ k)
+                             → X is k connected
+                             → X is l connected
+ connectedness-extends-below {𝓤} {X} {k} {l} o X-con =
+   connectedness-step-down (transport (λ z → X is z connected) p X-con)
+  where
+   m : ℕ
+   m = pr₁ (subtraction l k o)
+   p = k        ＝⟨ pr₂ (subtraction l k o) ⁻¹ ⟩
+       m +' l   ＝⟨ addition-commutativity m l ⟩
+       l +' m   ∎
+
+\end{code}
+
+We can characterize connected types in terms of inhabitedness and connectedness
+of the identity type at the level below. We will assume univalence when neccesary.
+
+\begin{code}
+
+ conn-to-inhabited : {X : 𝓤 ̇} {n : ℕ}
+                   → X is (succ n) connected
+                   → ∥ X ∥
+ conn-to-inhabited {_} {_} {0} X-conn =
+  center (equiv-to-singleton' one-trunc-≃-prop-trunc X-conn)
+ conn-to-inhabited {_} {_} {succ n} X-conn =
+  conn-to-inhabited (connectedness-is-lower-closed X-conn)
+
+ conn-to-id-conn : {X : 𝓤 ̇} {n : ℕ}
+                 → (ua : is-univalent 𝓤)
+                 → X is (succ n) connected
+                 → (x y : X) → (x ＝ y) is n connected
+ conn-to-id-conn {_} {_} {n} ua X-conn x y =
+  equiv-to-singleton (eliminated-trunc-identity-char ua)
+                     (is-prop-implies-is-prop' (singletons-are-props X-conn)
+                                               ∣ x ∣[ succ n ]
+                                               ∣ y ∣[ succ n ])
+
+ conn-to-inhabited-id-conn : {X : 𝓤 ̇} {n : ℕ}
+                           → (ua : is-univalent 𝓤)
+                           → X is (succ n) connected
+                           → ∥ X ∥ × ((x y : X) → (x ＝ y) is n connected)
+ conn-to-inhabited-id-conn ua X-conn =
+  (conn-to-inhabited X-conn , conn-to-id-conn ua X-conn)
+
+ inhabited-id-conn-to-conn : {X : 𝓤 ̇} {n : ℕ}
+                           → (ua : is-univalent 𝓤)
+                           → ∥ X ∥ × ((x y : X) → (x ＝ y) is n connected)
+                           → X is (succ n) connected
+ inhabited-id-conn-to-conn {_} {_} {0} ua (anon-x , id-conn) =
+  pointed-props-are-singletons (prop-trunc-to-one-trunc anon-x) one-trunc-is-prop
+ inhabited-id-conn-to-conn {_} {_} {succ n} ua (anon-x , id-conn) =
+  ∥∥-rec (being-singleton-is-prop fe)
+         (λ x → (∣ x ∣[ succ (succ n) ]
+          , ∥∥ₙ-ind (λ v → hlevels-are-upper-closed
+                            (λ p q → ∥∥ₙ-hlevel ∣ x ∣[ succ (succ n) ] v p q)) 
+                    (λ y → forth-trunc-id-char ua (center (id-conn x y)))))
+         anon-x
+
+ connected-characterization : {X : 𝓤 ̇} {n : ℕ}
+                            → (ua : is-univalent 𝓤)
+                            → X is (succ n) connected
+                            ↔ ∥ X ∥ × ((x y : X) → (x ＝ y) is n connected)
+ connected-characterization ua =
+  (conn-to-inhabited-id-conn ua , inhabited-id-conn-to-conn ua)
+
+ ap-is-less-connected : {X : 𝓤 ̇} {Y : 𝓥 ̇} {n : ℕ} {x x' : X}
```

Comment:
It's nicer in practice and conceptually cleaner to introduce `{x x' : X}` after introducing `f` and the fact that it's `n + 1`-connected. (That way you can forget about all the other implicit variables.)